src/adaptors/http2/http2_adaptor.h

```diff
@@ -138,6 +141,7 @@ struct qdr_http2_connection_t {
     bool                      ingress;
     bool                      timer_scheduled;
     bool                      client_magic_sent;
+    time_t                    prev_ping; // Time the previous PING frame was sent on egress connection.
```

Comment:
Strictly this could be the last time for any write to the socket, not just a ping. The time could then be updated on every write.

Reply:
The check_send_ping_frame() only conditionally updates the conn->prev_ping field like this - 



    time_t current = time(NULL);
    time_t prev = conn->prev_ping;
    if (current - prev >= 4) {
        send_ping_frame(conn);
        qd_log(http2_adaptor->log_source, QD_LOG_TRACE, "[C%"PRIu64"] Sent PING frame", conn->conn_id);
        qd_timer_schedule(conn->ping_timer, 4000);
        conn->prev_ping = current;
    }

So the conn->prev_ping contains the time that the ping was sent out ?

Reply:
Right, and that is fine. All I'm saying is that if you send out any other frame within the second, there is no need to send a ping. However that is an optimisation. It is good as is.

Reply:
Ah, yes, agreed. Good point.